### PR TITLE
enhancement/filter-rpc-tests

### PR DESCRIPTION
**Pull request by decanus**
Adds tests to validate the filter rpc behaviour.

What happens is the following:
 - Conn subscribes to filter using `topic` 
 - We send messages onto that topic.
 - User receives messages.

### CODE REVIEW

**Review by oskarth — APPROVED** [review]

File: tests/v2/test_waku_filter.nim [hunk 1]
```diff
@@ -18,3 +26,54 @@ procSuite "Waku Filter":
     check:
       decode.isErr == false
       decode.value == rpc
+
+  asyncTest "handle filter":
+    let
+      proto = WakuFilter.init()
```
Comment (oskarth) [inline tests/v2/test_waku_filter.nim:32]: Are we sending WakuFilter over the wire? I thought the only thing we are sending over the wire here is related to ContentFilter, or possibly some RPC.

If we aren't sending it over the wire, why is this a protobuf?

File: tests/v2/test_waku_filter.nim [hunk 2]
```diff
@@ -18,3 +26,54 @@ procSuite "Waku Filter":
...
+    let transport2: TcpTransport = TcpTransport.init()
+    let conn = await transport2.dial(transport1.ma)
+
+    var rpc = FilterRPC(filters: @[ContentFilter(topics: @["topic"])])
+    discard await msDial.select(conn, WakuFilterCodec)
+    await conn.writeLP(rpc.encode().buffer)
+
+    await sleepAsync(2.seconds)
+
+    filters.notify(msg)
```
Comment (oskarth) [inline tests/v2/test_waku_filter.nim:71]: Isn't this just local? Doesn't seem to touch the wire at all

**Comment by oskarth** [comment]
I'm not quite sure what is going on here and with filter protocol semantics, but seems fine I suppose

File: tests/v2/test_waku_filter.nim (hunk 2 above)
Comment (decanus) [reply tests/v2/test_waku_filter.nim:71]: Yes this is local, what it does is notify the filter protocol locally that we have a new message. This then takes it and sends it to the subscriber.

Comment (oskarth) [reply tests/v2/test_waku_filter.nim:71]: Seems to be complecting with relay protocol ("sends to subscriber" how?), but let's go ahead and merge and then discuss semantics in spec PR

Comment (kdeme) [reply tests/v2/test_waku_filter.nim:71]: Hmm, I had the same question. This notify should get triggered from an incoming message (or messages) to which we subscribed via contentFilters subscribe. But pushing messages is not in the filter protocol yet so this cannot be tested.

So all that could be tested is checking if the `ContentFilter` is being received well on the other side. But iiuc, this is not what is being done here. So I'm not sure if I correctly understand the point of this test right now.

Comment (decanus) [reply tests/v2/test_waku_filter.nim:71]: Yes the message is being pushed... the filter is being notified for example if a message is received by WakuRelay. Which we simulate in the test by calling filter.notify. The handler function we then have defined in the WakuFilter protocol checks for the received message, sees if any user has subscribed to the topics that the message is a part of and sends them out to that connection.

Comment (kdeme) [reply tests/v2/test_waku_filter.nim:71]: OK, so you are talking about the communication in the other direction. Getting messages received on the Gossip domain (waku relay) to be send to the Waku Filter subscribers, if they match the contentFilter.

But still, that is not what is being tested here, right? As that part of the RPC is not yet implemented?
I see that  in this test the actual `contentFilter` is being send (and checked for receival), which is a fine test. I just don't understand why the notify is required in that test.

Comment (decanus) [reply tests/v2/test_waku_filter.nim:71]: It most definately is being tested here.

we literally assert that we receive the message that was simulated on the server side:

```nim
var message = await conn.readLp(64*1024)
let response = protobuf.decodeMessage(initProtoBuffer(message))

check:
  msg == response.value
```

Comment (kdeme) [reply tests/v2/test_waku_filter.nim:71]: Right, I thought we were reading the response from the FilterRPC that got send.

**Pull request merged**